APRIORI

1. Import Library

In [67]:
!pip install apyori

In [68]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

2. Pra Proses Data

In [69]:
# Mounted At/Content/Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [70]:
dataset = pd.read_csv('/content/drive/MyDrive/KuliahUNSIA/Semester5/MachineLearning/Peretmuan16/Tugas2/Grocery_Store_DataSet.csv', header = None)
transactions = []
for i in range(0, 7501):
  transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])

In [71]:
print(dataset)

           0                    1                    2                 3   \
0     Item(s)               Item 1               Item 2            Item 3   
1           4         citrus fruit  semi-finished bread         margarine   
2           3       tropical fruit               yogurt            coffee   
3           1           whole milk                  NaN               NaN   
4           4            pip fruit               yogurt      cream cheese   
...       ...                  ...                  ...               ...   
9831       17              sausage              chicken              beef   
9832        1    cooking chocolate                  NaN               NaN   
9833       10              chicken         citrus fruit  other vegetables   
9834        4  semi-finished bread        bottled water              soda   
9835        5              chicken       tropical fruit  other vegetables   

                  4              5               6                7   \
0  

3. Menampilkan Sebanyak 5 Data Teratas

In [74]:
df= dataset = pd.read_csv('/content/drive/MyDrive/KuliahUNSIA/Semester5/MachineLearning/Peretmuan16/Tugas2/Grocery_Store_DataSet.csv', header = None)
df.head(5)

0               1                    2             3             4   \
0  Item(s)          Item 1               Item 2        Item 3        Item 4   
1        4    citrus fruit  semi-finished bread     margarine   ready soups   
2        3  tropical fruit               yogurt        coffee           NaN   
3        1      whole milk                  NaN           NaN           NaN   
4        4       pip fruit               yogurt  cream cheese  meat spreads   

       5       6       7       8       9   ...       23       24       25  \
0  Item 5  Item 6  Item 7  Item 8  Item 9  ...  Item 23  Item 24  Item 25   
1     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   
2     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   
3     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   
4     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   

        26       27       28       29       30       31       32  
0  Item 26  Item 27  Item 28  Item 29  Item 30  Item 31  Item 32  
1      NaN      NaN      NaN      NaN      NaN      NaN      NaN  
2      NaN      NaN      NaN      NaN      NaN      NaN      NaN  
3      NaN      NaN      NaN      NaN      NaN      NaN      NaN  
4      NaN      NaN      NaN      NaN      NaN      NaN      NaN  

[5 rows x 33 columns]

4. Melihat Variabel dalam Dataset

In [75]:
df.keys()

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32],
           dtype='int64')

5. Melihat Shape

In [76]:
df.shape

(9836, 33)

Di sini terlihat bahwa terdapat 9836 baris dan 33 kolom.

6. Training Apriori

In [77]:
from apyori import apriori
rules = apriori(transactions = transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2, max_length = 2)

7. Menampilkan Hasil Training

In [78]:
results = list(rules)
for rule in results:
  print(rule)

RelationRecord(items=frozenset({'10', 'fruit/vegetable juice'}), support=0.0061325156645780565, ordered_statistics=[OrderedStatistic(items_base=frozenset({'10'}), items_add=frozenset({'fruit/vegetable juice'}), confidence=0.25136612021857924, lift=3.46598762455802)])
RelationRecord(items=frozenset({'10', 'margarine'}), support=0.005465937874950006, ordered_statistics=[OrderedStatistic(items_base=frozenset({'10'}), items_add=frozenset({'margarine'}), confidence=0.2240437158469945, lift=3.734559805707346)])
RelationRecord(items=frozenset({'10', 'root vegetables'}), support=0.008132249033462205, ordered_statistics=[OrderedStatistic(items_base=frozenset({'10'}), items_add=frozenset({'root vegetables'}), confidence=0.3333333333333333, lift=3.0160836349014875)])
RelationRecord(items=frozenset({'domestic eggs', '11'}), support=0.0038661511798426876, ordered_statistics=[OrderedStatistic(items_base=frozenset({'11'}), items_add=frozenset({'domestic eggs'}), confidence=0.21804511278195488, lift=3

8. Merubah Dataset Ke Dalam Bentuk Kolom

In [79]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])

9. Melihat Hasil

In [80]:
print(resultsinDataFrame)

      Left Hand Side        Right Hand Side   Support  Confidence      Lift
0                 10  fruit/vegetable juice  0.006133    0.251366  3.465988
1                 10              margarine  0.005466    0.224044  3.734560
2                 10        root vegetables  0.008132    0.333333  3.016084
3                 11          domestic eggs  0.003866    0.218045  3.379249
4                 11              margarine  0.003866    0.218045  3.634570
5                 12          domestic eggs  0.003466    0.276596  4.286663
6                 12  fruit/vegetable juice  0.003200    0.255319  3.520494
7                 12       other vegetables  0.007599    0.606383  3.156474
8                 12              pip fruit  0.003333    0.265957  3.445504
9                 12        root vegetables  0.004399    0.351064  3.176514
10                12         tropical fruit  0.004133    0.329787  3.139256
11                12     whipped/sour cream  0.003333    0.265957  3.640414
12          

10. Kesimpulan

Berdasarkan hasil perhitungan menggunakan python di atas dapat disimpulkan bahwa: pembeli yang membeli baking powder
 (bubuk pengembang) kemungkinan akan membeli sugar (gula), pembeli yang membeli baking powder (bubuk pengembang) kemungkinan juga akan membeli  (krim kue), pembeli yang membeli beef (daging sapi) kemungkinan akan membeli root vegetables (sayuran berakar), pembeli yang membeli berries (buah beri) kemungkinan akan membeli whipped/sour cream (krim kue), pembeli yang membeli liquor (minuman beralkohol) kemungkinan akan membeli bootled beer (bir botol), pembeli yang membeli red/blushwine (anggur merah) kemungkinan akan membeli bottled beer (bir botol), pembeli yang membeli flour (tepung) kemungkinan akan membeli margarine (mentega), pembeli yang membeli flour (tepung) kemungkinan akan membeli sugar (gula), pembeli yang membeli flour (tepung) kemungkinan akan membeli whipped/sour cream (krim kue), pembeli yang membeli mustard (saus moster) kemungkinan akan membeli frankfurter (sosis), pembeli yang membeli processed cheese (keju) kemungkinan akan membeli fruit/vegetable juice (jus sayur/buah), pembeli yang membeli grapes (anggur) kemungkinan akan membeli tropical fruit (buah tropis), pembeli yang membeli processed cheese (keju) kemungkinan akan membeli ham (daging babi yang dikeringkan), pembeli yang membeli ham (daging babi yang dikeringkan) kemungkinan akan membeli white bread (roti putih), pembeli yang membeli herbs (tanaman herba) kemungkinan akan membeli root vagetables (sayuran akar), pembeli yang membeli rice (beras) kemungkinan akan membeli root vegetables (sayuran berakar), pembeli yang membeli sliced cheese (irisan keju) kemungkinan akan membeli sausage (sosis), dan pembeli yang membeli turkey (ayam kalkun) kemungkinan akan membeli tropical fruit (buah tropis).

Terima kasih